core/vm/evm.go

```diff
@@ -220,6 +227,32 @@ func NewEVM(blockCtx BlockContext, txCtx TxContext, statedb StateDB, chainConfig
 		chainRules:  chainConfig.Rules(blockCtx.BlockNumber, blockCtx.Time),
 	}
 	evm.interpreter = NewEVMInterpreter(evm)
+
+	// If the predicate results are already available, use them.
+	if blockCtx.PredicateResults != nil {
+		return evm
+	}
+
+	// Parse the predicate results from the extra field and store them in the
+	// block context.
+	predicateBytes := header.PredicateBytesFromExtra(blockCtx.Extra)
```

Comment:
Why is the `Extra []byte` being copied from `Header` to `BlockContext` if it's eventually being interpreted by a method on the former? This suggests that either the copy is unnecessary (if the slices are the same) or that having the method defined on the `Header` is inappropriate (if the slices are different) because the two objects have become decoupled. I've not done much work in this predicate stuff so I may not have context, but this feels like a code smell that could point to something more serious.

Reply:
header is a package here